File: Travelling_Salesman_Problem/Travelling_Salesman.java
//Travelling Salesman Problem

import java.util.*;
class TSP
{
    // Function to find the minimum weight Hamiltonian Cycle
    
    static int tspfn( int[][] graph, boolean[] visited, int currPosition, int n, int count, int cost, int answer)
     {
        //If last node is reached and it has a link to the starting node i.e the source then
       //keep the minimum value out of the total cost of traversal and "answer"
      //Finally return to check for more possible values
      
      if( count == n && graph[currPosition][0] > 0 )
      {
        answer = Math.min( answer, cost + graph[currPosition][0]);
        return answer;
      }

      //BACKTRACKING STEP
     //Loop to traverse the adjacency list of currPos node and increasing the count  by 1 and cost by graph[currPos,i] value
     
     for( int i = 0; i < n; i++)
     { 
      if( visited[i] == false && graph[currPosition][i] > 0)
      {
       //Mark as visited
       visited[i] = true;
       answer = tspfn( graph, visited, i, n, count + 1, cost + graph[currPosition][i], answer);
       //Mark ith node as unvisited
       visited[i] = false;
      }
    }
     return answer;
   }
   
    public static void main( String[] args)
    {
       //n is the number of nodes i.e. V
       int n;
       System.out.println(" Enter the number of villages: ");
       Scanner in = new Scanner( System.in );
       n = in.nextInt();
       int graph[][] = new int[n][n];
       System.out.println( "Enter the cost matrix " );
       for( int i=0; i < n; i++)
       {
         System.out.println( "Enter the cost matrix for village " + i+1 );
         for( int j=0; j<n; j++)
         graph[i][j] = in.nextInt();
       }
       
       //Boolean array to check if a node has been visited or not
       boolean[] visited = new boolean[n];

      //Mark 0th node as visited
      visited[0] = true;
      int answer = Integer.MAX_VALUE;

     //Find the minimum weight Hamiltonian Cycle
     answer = tspfn( graph, visited, 0, n, 1, 0, answer);
     
     //answer is the minimum weight Hamiltonian Cycle
     System.out.println(" The minimum cost is " +answer);
  }
}

/* Sample input
Enter the number of villages:
0
Enter the cost matrix
Enter the cost matrix for village 1
0 10 15 20
Enter the cost matrix for village 2
10 0 35 25
Enter the cost matrix for village 3
15 35 0 30
Enter the cost matrix for village 4
20 25 30 0

Sample Output
The minimum cost is 80 */

File: Travelling_Salesman_Problem/Travelling_Salesman.cpp
#include<iostream>
using namespace std;

int ary[10][10], completed[10], n, cost = 0;

void takeInput()
{
    int i, j;
 
    cout << "Enter the number of villages: ";
    cin >> n;
 	
    //Enter distance between cities
    cout << "\n Enter the Cost Matrix \n";
 
    for (i = 0; i < n; i++)
    {
        cout << "\n Enter Elements of Row " << i+1 << "\n";
 
        for (j = 0; j < n; j++)
            cin >> ary[i][j];
 
        completed[i] = 0;
    }
 	
    //Display matrix of distance between cities
    cout << "\n\n The cost list is:";
 
    for (i = 0; i < n; i++)
    {
        cout << "\n";
 
        for (j = 0; j < n; j++)
            cout << "\t" << ary[i][j];
    }
}

//Function for TSP algo
int least(int c)
{
    int i, nc = 999;
    int min = 999, kmin;
  
    for (i = 0; i < n; i++)
    {
        if ((ary[c][i] != 0) && (completed[i] == 0))
            if (ary[c][i] + ary[i][c] < min)
            {
                min = ary[i][0] + ary[c][i];
                kmin = ary[c][i];
                nc = i;
            }
    }
 
    if (min != 999)
	    cost += kmin;
 
    return nc;
}
 
//Function to determine minimum cost
void mincost(int city)
{
    int i, ncity;
 
    completed[city] = 1;
 
    cout << city + 1 << "--->";
    ncity = least(city);
 
    if (ncity == 999)
    {
        ncity = 0;
        cout << ncity + 1;
        cost += ary[city][ncity];
 
        return;
    }
 
    mincost(ncity);
}
 
//Display results
int main()
{
    takeInput();
 
    cout << "\n\n The Path is: \n";
    mincost(0); //passing 0 because starting vertex
 
    cout << "\n\n Minimum cost is " << cost;
 
    return 0;
}

/* Output:

Enter the number of villages: 4                                                                                         
                                                                                                                        
Enter the Cost Matrix                                                                                                   
                                                                                                                        
Enter Elements of Row: 1                                                                                                
0 10 15 20 

Enter Elements of Row 2                                                                                                 
30 40 15 0                                                                                                              
                                                                                                                        
Enter Elements of Row 3                                                                                                 
0 0 35 25  

Enter Elements of Row 4                                                                                                 
20 10 0 30                                                                                                              
                                                                                                                        
                                                                                                                        
The cost list is:                                                                                                       
        10      15      0       20                                                                                      
        30      40      15      0 
        0       0       35      25                                                                                      
        20      10      0       30                                                                                      
                                                                                                                        
The Path is:                                                                                                            
1--->4--->2--->3--->1 

Minimum cost is 45                                                                                                  
*/
